1.14.x -> 1.15.x Update

ChampionAsh5357 opened the pull request:
This PR is an overhaul of #313 into 1.14.x as specified by tterag1098. About 90% of the changes are present here. There are very few differences like some formatting fixes. The only major thing of note is commenting out Animation and WorldSavedData within the navigation bar due to its lack of updated information.

The original PR in #322 was pulled incorrectly. Since then, there have been a number of updates including fixes in BlockStates for properties. Any updated changes has been removed and little to no cleanup has been done on those files.

Please feel free to let me know if there is anything I need to add or fix from transferring over.

ChampionAsh5357 committed:
Corrected rebranching of 1.14.x

sciwhiz12 commented on docs/blocks/states.md:
```diff
@@ -61,7 +61,7 @@ In your Block class, create or reference `static final` `IProperty<?>` objects f
   * `DirectionProperty`
     * This is a convenience implementation of `EnumProperty<Direction>`
-    * Several convenience predicates are also provided. For example, to get a property that represents the cardinal directions, call `DirectionProperty.create("<name>", Direction.Plane.HORIZONTAL)`; to get the X directions, `DirectionProperty.create("<name>", Direction.Axis.X)`
+    * Several convenience predicates are also provided. For example, to get a property that represents the cardinal directions, call `DirectionProperty.create("<name>", Direction.Plane.HORIZONTAL)` to get the X directions, `DirectionProperty.create("<name>", Direction.Axis.X)`
```
The semicolon is intentional; it separates the first clause from the second clause.
> to get a property that represents the cardinal directions, call `DirectionProperty.create("<name>", Direction.Plane.HORIZONTAL)`
> to get the X directions, `DirectionProperty.create("<name>", Direction.Axis.X)

ChampionAsh5357 replied:
Ah apologies. I remember that there was a lot of formatting discrepancies in the original and so I assumed without reading that the semicolon was a leftover of that.

ChampionAsh5357 committed:
Added back the semicolon

ChampionAsh5357 committed:
Little update to capabilities

Appends information associated with itemstacks, the default implementation method, and the invalidation of optionals.

All other information that was mentioned is already in the docs.

ChampionAsh5357 commented:
@tterrag1098 If you wouldn't mind look this over so the docs can be fully up to date.

sciwhiz12 commented on docs/animation/implementing.md:
```diff
@@ -8,21 +8,21 @@ you are animating.
 --------
 
-Animations for blocks are done with the `AnimationTESR`, which is a `FastTESR`. Because of this, having a `TileEntity` for your block
+Animations for blocks are done with the `TileEntityRendererAnimation`, which is a `TileEntityRendererFast`. Because of this, having a `TileEntity` for your block
```
> Animations for blocks are done with the `TileEntityRendererAnimation`, which is a `TileEntityRendererFast`. Because of this, having a `TileEntity` for your block

Seems to be a bit outdated, since it does not extend `TileEntityRendererFast` (and no such class exists), but instead `TileEntityRenderer` itself.

On the topic of the animations docs, should all of the Animation State Machine API docs still be kept? Someone needs to test if it still works, or check if it's infeasible to fix it to work and document.

ChampionAsh5357 replied:
Last time I checked, the animation API does not work as of 1.15 or was scheduled for removal. Hence, if you check the sidebar, I have that commented out since the docs pages would not be relevant anymore.

sciwhiz12 commented on docs/blocks/blocks.md:
```diff
@@ -34,7 +34,7 @@ Blocks must be [registered][registering] to function.
     A block in the world and a "block" in an inventory are very different things. A block in the world is represented by an `BlockState`, and its behavior defined by an instance of `Block`. Meanwhile, an item in an inventory is an `ItemStack`, controlled by an `Item`. As a bridge between the different worlds of `Block` and `Item`, there exists the class `BlockItem`. `BlockItem` is a subclass of ` ...
 
-    When a block is registered, *only* a block is registered. The block does not automatically have an `BlockItem`. To create a basic `BlockItem` for a block, one should use `new BlockItem(block).setRegistryName(block.getRegistryName())`. Custom subclasses of `BlockItem` may be used as well. Once an `BlockItem` has been registered for a block, `Item.getItemFromBlock` can be used to retrieve it. ` ...
+    When a block is registered, *only* a block is registered. The block does not automatically have an `BlockItem`. To create a basic `BlockItem` for a block, one should use `new BlockItem(block).setRegistryName(block.getRegistryName())`. Custom subclasses of `BlockItem` may be used as well. Once an `BlockItem` has been registered for a block, `Block#asItem` can be used to retrieve it. `Block#asI ...
```
> `Block#asItem` will return `null` if there is no `BlockItem` for the `Block`, so if you are not certain that there is an `BlockItem` for the `Block` you are using, check for `null`.

Looking at the 1.15.2 code, it seems that `Block#asItem` will never return `null` because `Item.getItemForBlock(Block)` is used, which returns `Items.AIR` if there is no block (and it should be impossible for the returned object/referrent from the registry delegate to be `null`).

sciwhiz12 commented on docs/blocks/interaction.md:
```diff
@@ -108,7 +108,7 @@ Called on a block when it is clicked by a player.
+Two blocks that override this method are the `NoteBlock` and the `RedstoneOreBlock`.
 
 The Note block overrides this method so that when left-clicked, it plays a sound.  
 The RedstoneOre block overrides method so that when left-clicked, it gives off emits faint light for a few seconds.
```
The following paragraph should also be fixed. And could you also consolidate the examples into one neater paragraph instead?

sciwhiz12 commented on docs/blocks/interaction.md:
````diff
@@ -138,9 +138,9 @@ This method has important behavior in the `Block` class so be sure to call the s
 ```
 
-The **TNT Block** overrides this method to cause it's explosion when a player destroys it.  
+The `TNTBlock` overrides this method to cause it's explosion when a player destroys it.  
````
This needs to be clarified, that the TNT block only is ignited by being harvested/destroyed when the `unstable` state is `true`.

sciwhiz12 commented on docs/datastorage/capabilities.md:
```diff
@@ -32,57 +32,59 @@ static Capability<IItemHandler> ITEM_HANDLER_CAPABILITY = null;
 The second method can be used to provide custom implementations. In the case of `IItemHandler`, the default implementation uses the `ItemStackHandler` class, which has an optional argument in the constructor, to specify a number of slots. However, relying on the existence of these default implementations should be avoided, as the purpose of the capability system is to prevent loading errors in co ...
 
-Once you have your own instance of the capability interface, you will want to notify users of the capability system that you expose this capability. This is done by overriding the `hasCapability` method, and comparing the instance with the capability you are exposing. If your machine has different slots based on which side is being queried, you can test this with the `facing` parameter. For Entit ...
+Once you have your own instance of the capability interface, you will want to notify users of the capability system that you expose this capability and provide an optional of the interface reference. This is done by overriding the `getCapability` method, and comparing the instance with the capability you are exposing. If your machine has different slots based on which side is being queried, you c ...
```
>(top side => head slot?)

> (west => slot on the left?)

The question mark makes it look like more of a guess. Could you reword it, like `(e.g. west for the left side)`?

> Make sure to invalidate the optional of the reference instance at the end of the provider's lifecycle.

Might consider moving this to a seperate paragraph, and clarifying more when a `LazyOptional` should be invalidated (such as TE removal, or change of the state like slot changes).

sciwhiz12 commented on docs/models/blockstates/introduction.md:
```diff
@@ -3,29 +3,26 @@ Introduction to Blockstate JSONs
-    "axis=none":   { "model": "oak_bark" }
-  }
+    "variants": {
+        "axis=y":    { "model": "block/oak_log" },
```
Just a small formatting matter, this should be in line with the other two cases.

sciwhiz12 commented on docs/models/color.md:
```diff
@@ -5,15 +5,29 @@ Many blocks and items in vanilla change their texture color depending on where t
 ### `IBlockColor`/`IItemColor`
 
-Both of these are single-method interfaces. `IBlockColor` takes an `IBlockState`, an (nullable) `ILightReader`, and a (nullable) `BlockPos`. `IItemColor` takes an `ItemStack`. Both of them take a parameter `tintindex`, which is the tint index of the face being colored. Both of them return an `int`, a color multiplier. This `int` is treated as 4 unsigned bytes, alpha, red, green, and blue, in that ...
+Both of these are single-method interfaces. `IBlockColor` takes an `IBlockState`, an (nullable) `ILightReader`, and a (nullable) `BlockPos`. `IItemColor` takes an `ItemStack`. Both of them take a parameter `tintIndex`, which is the tint index of the face being colored. Both of them return an `int`, a color multiplier. This `int` is treated as 4 unsigned bytes, alpha, red, green, and blue, in that ...
```
> **...** `IBlockColor` takes an `IBlockState`, an (nullable) **...***

`IBlockState` should be updated to `BlockState`.

sciwhiz12 commented on docs/tileentities/tesr.md:
```diff
@@ -25,4 +25,4 @@ This method is called every frame in order to render the tile entity.
 -----------------
 
-In order to register a TESR, call `ClientRegistry#bindTileEntitySpecialRenderer` passing the tile entity class to be renderer with this TER and the instance of the TER to use to render all TEs of this class.
+In order to register a TER, call `ClientRegistry#bindTileEntitySpecialRenderer` passing the tile entity class to be renderer with this TER and the instance of the TER to use to render all TEs of this class.
```
The method is now named `ClientRegistry#bindTileEntityRenderer`.

sciwhiz12 commented on docs/tileentities/tileentity.md:
````diff
@@ -125,7 +125,7 @@ as you can make sure that only the data you need to be synchronized is actually
 Once you've created your custom network message, you can send it to all users that have the `TileEntity` loaded with:
 ```JAVA
-SimpleNetworkWrapper#sendToAllTracking(IMessage, NetworkRegistry.TargetPoint)
+SimpleChannel#send(PacketDistributor.PacketTarget, MSG)
````
Should this be in a code fence? Looks like it should be an inline code block.

sciwhiz12 reviewed (changes requested):
Needs some changes, a few things were skipped over. After that, should be good for merging.

ChampionAsh5357 committed:
Address typos and misleading/unclear sentences

tterrag1098 commented on docs/datastorage/capabilities.md:
```diff
@@ -32,57 +32,61 @@ static Capability<IItemHandler> ITEM_HANDLER_CAPABILITY = null;
+LazyOptional<IItemHandler> inventoryHandlerLazyOptional;
+
+// After initializing inventoryHandler
+inventoryHandlerLazyOptional = LazyOptional.of(() -> inventoryHandler);
```
I don't agree with this example, the point of *laziness* is to initialize the result only when needed.

tterrag1098 commented on docs/datastorage/capabilities.md:
```diff
@@ -32,57 +32,61 @@ static Capability<IItemHandler> ITEM_HANDLER_CAPABILITY = null;
-  }
-  return super.getCapability(capability, facing);
+public void remove() {
+	super.remove();
```
This indentation is inconsistent

tterrag1098 commented on docs/models/advanced/ibakedmodel.md:
```diff
@@ -38,16 +38,14 @@ See [Perspective][].
 
 ### `getQuads`
 
-This is the main method of `IBakedModel`. It returns `BakedQuad`s, which contain the low-level vertex data that will be used to render the model. If the model is being rendered as a block, then the `IBlockState` passed in is non-null. Additionally, [`Block::getExtendedState`][extended blockstates] is called to create the passed `IBlockState`, which allows for arbitrary data to be passed from the  ...
```
If we're removing the reference to getExtendedState it should be replaced with at least a brief mention of IModelData

tterrag1098 commented on docs/models/blockstates/introduction.md:
```diff
@@ -3,29 +3,26 @@ Introduction to Blockstate JSONs
-A blockstate always has to be defined for all possible variant strings. When you have many properties, this results in lots of possible variants, as every combination of properties must be defined. In Minecraft 1.8's blockstate format, you have to define every string explicitly, which leads to long, complicated files. It also doesn't support the concept of submodels, or multiple models in the sam ...
-
-Starting from Minecraft 1.9, Mojang also introduced the "multipart" format. You can find a definition of its format on the [wiki][]. Forge's format and the multipart format are not better than each other; they each cover different use cases and it is your choice which one you want to use.
+A blockstate always has to be defined for all possible variant strings. When you have many properties, this results in lots of possible variants, as every combination of properties must be defined. In Minecraft 1.8's blockstate format, you have to define every string explicitly, which leads to long, complicated files. Starting from Minecraft 1.9, Mojang also introduced the "multipart" format. You ...
```
This is not correct.

tterrag1098 commented on docs/models/blockstates/introduction.md:
```diff
@@ -3,29 +3,26 @@ Introduction to Blockstate JSONs
 
 !!! note
-    The Forge format is really more like syntactic sugar for automatically calculating the set of all possible variants for you behind the scenes. This allows you to use the resulting `ModelResourceLocation`s for things other than blocks. ([Such as items][item blockstates]. This is also true of the 1.8 format, but there is almost no reason to use that format.) The 1.9 format is a more complicated ...
+    The Forge format is really more like syntactic sugar for automatically calculating the set of all possible variants for you behind the scenes. This allows you to use the resulting `ModelResourceLocation`s for things other than blocks. The 1.9 format is a more complicated system that depends on having an `IBlockState` to pick the model. It will not directly work in other contexts without some  ...
```
This note should be removed entirely.

tterrag1098 commented on docs/models/blockstates/introduction.md:
```diff
@@ -3,29 +3,26 @@ Introduction to Blockstate JSONs
-    The Forge format is really more like syntactic sugar for automatically calculating the set of all possible variants for you behind the scenes. This allows you to use the resulting `ModelResourceLocation`s for things other than blocks. ([Such as items][item blockstates]. This is also true of the 1.8 format, but there is almost no reason to use that format.) The 1.9 format is a more complicated ...
+    The Forge format is really more like syntactic sugar for automatically calculating the set of all possible variants for you behind the scenes. This allows you to use the resulting `ModelResourceLocation`s for things other than blocks. The 1.9 format is a more complicated system that depends on having an `IBlockState` to pick the model. It will not directly work in other contexts without some  ...
 
 For reference, here's an excerpt from the 1.8 blockstate for fences, `fence.json`:
```
Most of this line and below is incorrect

tterrag1098 commented on docs/tileentities/tileentity.md:
````diff
@@ -81,7 +81,7 @@ TileEntity#getUpdateTag()
 ```
 Again, this is pretty simple, the first method collects the data that should be send to the client,
-while the second one processes that data. If your `TileEntity` doesn't contain much data you might be able to use the methods out of the `Storing Data within your TileEntity` section.
+while the second one processes that data. If your `TileEntity` doesn't contain much data you might be able to use the methods out of the [Storing Data within your TileEntity][storing-data] section.
````
```suggestion
while the second one processes that data. If your `TileEntity` doesn't contain much data you might be able to use the methods out of the [Storing Data within your `TileEntity`][storing-data] section.
```

tterrag1098 commented:
These changes are obviously quite important, sorry for the extreme delay. Most of my review notes are minor, the main thing is the outdated info about (forge) blockstates.

ChampionAsh5357 committed:
Fix BlockState JSONs, Formatting, and IModelData

ChampionAsh5357 committed:
Merge remote-tracking branch 'upstream/1.15.x' into 1.14.x-new

ChampionAsh5357 commented:
Should be all good now. Apologies for the errors. The original push was just an update of the text and mappings rather than a rewriting of semi-outdated information.

tterrag1098 commented on docs/models/blockstates/introduction.md:
```diff
@@ -19,28 +19,29 @@ As an example, let's take a look at the vanilla `oak_log.json`:
 Here we define 3 variant strings, and for each we use a certain model, either the upright log and the sideways log (rotated in the y direction or not). These variants will define the look of a log depending on the property `axis`.
 
-A blockstate always has to be defined for all possible variant strings. When you have many properties, this results in lots of possible variants, as every combination of properties must be defined. In Minecraft 1.8's blockstate format, you have to define every string explicitly, which leads to long, complicated files. Starting from Minecraft 1.9, Mojang also introduced the "multipart" format. You ...
+A blockstate must be defined for all possible variants that invoke a change in the current model. Any property not specified in the JSON will not have any bearing to determine the current model (e.g. `waterlogged` has no effect on how a model might look). As each property adds a factor of two to the amount of states, there can be a number of possible variants needing a defined model.
```
I dislike that first sentence still, it contradicts itself just after. If you *must* define all variants, how can you not specify them as well? It's also mixing terms a bit, the individual things inside the blockstate are called *variants*, so calling them blockstates here is strange.

Additionally, each property does not add a factor of two, it depends how many valid values that property has. That last sentence is probably just unnecessary.

ChampionAsh5357 replied:
I'll reword that then quickly. My change in that sentence was that _must_ points to that sentence end where for all possible variants that invoke a change in the current model. It would be more applicable to say property, but I'll change the sentence such that it clarifies the point more.

ChampionAsh5357 committed:
Adjusts for some properties instead of all

sciwhiz12 commented on docs/models/advanced/ibakedmodel.md:
```diff
@@ -38,15 +38,16 @@ See [Perspective][].
 [Perspective]: perspective.md
 [ItemOverrideList]: itemoverridelist.md
-[ister]: /rendering/ister.md
+[ister]: ../../rendering/ister.md
```
Just to note, a previous PR was pulled that changed this to `/rendering/ister.md` which broke the link, because it made the hyperlink to `https://mcforge.readthedocs.io/rendering/ister.md`.

sciwhiz12 reviewed (approved):
Seems good to me, all of tterrag's reviews have been addressed.
